lapce-data/src/hover.rs

```diff
@@ -152,41 +170,316 @@ impl Default for HoverData {
+            | Options::ENABLE_TASKLISTS
+            | Options::ENABLE_HEADING_ATTRIBUTES,
+    );
+    for event in parser {
+        match event {
+            Event::Start(tag) => {
+                tag_stack.push((pos, tag));
+            }
+            Event::End(end_tag) => {
+                if let Some((start_offset, tag)) = tag_stack.pop() {
+                    if end_tag != tag {
+                        log::warn!("Mismatched markdown tag");
+                        continue;
+                    }
+
+                    if let Tag::CodeBlock(_kind) = &tag {
+                        code_block_indices.push(start_offset..pos);
+                    }
+
+                    add_attribute_for_tag(
+                        &tag,
+                        builder.add_attributes_for_range(start_offset..pos),
+                        style.clone(),
+                    );
+
+                    if should_add_newline_after_tag(&tag) {
+                        builder.push("\n");
+                        pos += 1;
+                    }
+                } else {
+                    // Very minor warning. We got a bit of malformed markdown (according to
+                    // pulldown_cmark, at least)
+                    log::warn!("Unbalanced markdown tag")
+                }
+            }
+            Event::Text(text) => {
+                builder.push(&text);
+                pos += text.len();
+            }
+            Event::Code(text) => {
+                builder.push(&text).font_family(FontFamily::MONOSPACE);
+                code_block_indices.push(pos..(pos + text.len()));
+                pos += text.len();
+            }
+            // TODO: Some minimal 'parsing' of html could be useful here, since some things use
+            // basic html like `<code>text</code>`.
+            Event::Html(text) => {
+                builder
+                    .push(&text)
+                    .font_family(FontFamily::MONOSPACE)
+                    .text_color(BLOCKQUOTE_COLOR);
+                pos += text.len();
+            }
+            Event::HardBreak => {
+                builder.push("\n");
+                pos += 1;
+            }
+            Event::SoftBreak => {
+                builder.push(" ");
+                pos += 1;
+            }
+            Event::Rule => {}
+            Event::FootnoteReference(_text) => {}
+            Event::TaskListMarker(_text) => {}
+        }
+    }
+
+    MarkdownText::new(builder.build(), code_block_indices)
+}
+
+fn add_attribute_for_tag(
+    tag: &Tag,
+    mut attrs: AttributesAdder,
+    HoverTextStyle {
+        base_font_size,
+        link_color,
+    }: HoverTextStyle,
+) {
+    use pulldown_cmark::HeadingLevel;
+    match tag {
+        Tag::Heading(level, _, _) => {
+            // The size calculations are based on the em values given at
+            // https://drafts.csswg.org/css2/#html-stylesheet
+            let font_size = match level {
```

Comment:
This might be a bit nicer if you returned only the scale factor here, and calculate the font size separately.